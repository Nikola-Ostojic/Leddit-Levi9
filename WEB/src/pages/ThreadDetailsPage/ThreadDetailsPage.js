import React, { useEffect, useState } from "react";
import { Container, LinearProgress, Card } from "@material-ui/core";
import { withRouter } from "react-router-dom";

import CreateComment from "components/CreateComment";
import { apiCall } from "utils/ApiCall";
import CommentItem from "components/CommentItem";
import CommentsFilter from "components/CommentsFilter";

import styles from "./ThreadDetailsPage.module.css";

const ThreadDetailsPage = (props) => {
  const { match, error, user } = props;

  const id = match?.params?.id;
  const [thread, setThread] = useState(null);
  const [errorThread, setErrorThread] = useState(null);

  const [comments, setComments] = useState([]);
  const [errorComment, setErrorComment] = useState(null);

  const [itemsPerPage, setItemsPerPage] = useState(5);
  const [currentPage, setCurrentPage] = useState(comments.page || 1);

  useEffect(() => {
    apiCall({
      url: `/threads/${id}`,
      method: "GET"
    })
      .then((result) => {
        const { data } = result;
        setThread(data);
      })
      .catch((e) => {
        setErrorThread(e.message);
      });

    fetchComments();
  }, [id, currentPage, itemsPerPage]);

  const fetchComments = () => {
    apiCall({
      url: `comments?threadId=${id}&page=${currentPage}&itemsPerPage=${itemsPerPage}`,
      method: "GET"
    })
      .then((result) => {
        const { data } = result;
        setComments(data);
      })
      .catch((e) => {
        setErrorComment(e.message);
      });
  };

  useEffect(() => {
    setCurrentPage(1);
  }, [itemsPerPage]);

  if (error) {
    return <p>{error}</p>;
  }

  if (errorThread) {
    return <div>{errorThread}</div>;
  }

  if (!thread && !errorThread) {
    return <LinearProgress />;
  }

  if (errorComment) {
    return <div>{errorComment}</div>;
  }
  if (!comments && !errorComment) {
    return <LinearProgress />;
  }

  return (
    <div>
      <Container>
        <Card className={styles.container}>
          <div className={styles.headerContainer}>
            <h2>{thread.title}</h2>
            <h2>{thread.author}</h2>
          </div>
          <div className={styles.content}>{thread.content}</div>
        </Card>
        <br />
        <CommentsFilter
          commentsCount={thread.commentsCount}
          totalPages={comments.totalPages}
          currentPage={currentPage}
          setCurrentPage={setCurrentPage}
          itemsPerPage={itemsPerPage}
          setItemsPerPage={setItemsPerPage}
        />
        {comments.data?.map((c) => (
          <CommentItem key={c.id} comment={c} />
        ))}
        <br />
        <br />
        {user && (
          <CreateComment threadId={id} onThreadCreated={fetchComments} />
        )}
      </Container>
    </div>
  );
};

export default withRouter(ThreadDetailsPage);
