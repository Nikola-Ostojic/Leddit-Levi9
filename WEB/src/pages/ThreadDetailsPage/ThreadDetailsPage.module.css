.container {
  padding: 15px;
  box-sizing: border-box;
  background-color: rgb(230, 230, 250);
}

.headerContainer {
  width: 100%;
  display: flex;
  justify-content: space-between;
}

.content {
  margin-top: 20px;
  margin-bottom: 20px;
  margin-right: 10px;
  color: darkslateblue;
}

.comments {
  padding: 15px;
}

body {
  background-color: rgb(230, 230, 250);
}
