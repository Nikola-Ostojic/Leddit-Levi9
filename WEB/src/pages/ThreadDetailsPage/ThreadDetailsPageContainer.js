import { connect } from "react-redux";

import { getUserRole, getUsername } from "modules/Auth/AuthSelectors";

import ThreadDetailsPage from "./ThreadDetailsPage";

export const mapStateToProps = (state) => ({
  role: getUserRole(state),
  user: getUsername(state)
});

export default connect(mapStateToProps)(ThreadDetailsPage);
