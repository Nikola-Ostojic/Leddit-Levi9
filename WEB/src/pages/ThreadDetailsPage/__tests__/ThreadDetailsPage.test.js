import React from "react";
import { shallow } from "enzyme";

import ThreadDetailsPage from "../ThreadDetailsPage";

describe("ThreadDetailsPage tests", () => {
  it("Should match snapshot", () => {
    const wrapper = shallow(<ThreadDetailsPage />);

    expect(wrapper).toMatchSnapshot();
  });
});
