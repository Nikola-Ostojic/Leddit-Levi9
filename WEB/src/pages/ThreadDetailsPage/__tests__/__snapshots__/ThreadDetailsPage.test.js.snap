// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreadDetailsPage tests Should match snapshot 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
